src/lib/y2storage/initial_guided_proposal.rb

```diff
@@ -38,7 +38,7 @@ class InitialGuidedProposal < GuidedProposal
     def initialize(settings: nil, devicegraph: nil, disk_analyzer: nil)
       super
 
-      @initial_settings = Yast.deep_copy(self.settings)
+      @initial_settings = self.settings.dup
```

Comment:
@lslezak,

Exactly, it "does not work". I guess you mean that we only need to use `initial_settings.dup` inside the `#reset_settings` method, since it is there where the copy is being made. But for the same reason we do not need to use none of them at this point, `Yast.deep_copy` nor `ProposalSettings#dup`. However, I think we want to hold a _real_ deep copy here.

@joseivanlopez, thoughts?

Reply:
>  However, I think we want to hold a real deep copy **of the initial/original settings** here.

Reply:
OK, then we should just add a comment why we are using own `deep_copy` implementation here as it might be confusing later.

Reply:
Just curiosity,  

Why do you think it is more confusing to see `self.settings.dup` than `Yast.deep_copy(self.settings)`?

In both case, what comes to my mind when reading the code is

> Ok, let's store a deep copy of current settings as `initial_settings`.

Reply:
Ah, I was not talking about using deep copy, that's fine regardless which implementation is used. 

I meant why we need implementing own version of deep copy and not using `Yast.deep_copy` directly. Something like

```ruby
# we cannot use Yast.deep_copy here because it keeps references to nested objects
```

(or something similar, I do not know the exact reason).

Reply:
@lslezak Well, that was exactly the root of the problem. @joseivanlopez  (who wrote the original code) assumed `Yast.deep_copy` would perform a deep copy of the passed object. But guess what: it does not!

`Yast.deep_copy` is quite a fail that only copies one level, leaving the second level as references instead of copies. It did its job with the YCP -> Ruby translation, but it should never be used in new code because it does not do what it promises. We should use Ruby mechanisms for performing deep copies, such as `Marshall`.

Reply:
Indeed, this copy could be optional because `#reset_settings` is call at the beginning of each try.

Reply:
> Indeed, this copy could be optional because #reset_settings is call at the beginning of each try.

So, what do you think it is better? Keep it as it is, or remove this copy?

Reply:
BTW, thanks to this discussion I realized that there is another place where we must to replace `Yast.deep_copy` by `ProposalSettings#(dup|deep_copy)`

`SettingsGenerator` has also an `initial_settings`

https://github.com/yast/yast-storage-ng/blob/06c72586a788e209bb04dc350a0a2ec66ba542b2/src/lib/y2storage/proposal/settings_generator/base.rb#L58-L60

Which is private method

https://github.com/yast/yast-storage-ng/blob/06c72586a788e209bb04dc350a0a2ec66ba542b2/src/lib/y2storage/proposal/settings_generator/base.rb#L90-L101

That make a "deep_copy" of the settings (ProposalSettings)

https://github.com/yast/yast-storage-ng/blob/06c72586a788e209bb04dc350a0a2ec66ba542b2/src/lib/y2storage/proposal/settings_generator/base.rb#L106-L109

So, we could change the `#initial_settings` to 

```ruby
def initial_settings
   return nil if @used_initial_settings
   @used_initial_settings = true

   settings.dup # or settings.deep_copy
end
```

Right?

Reply:
Something similar for `SettingsGenerator#initialize_settings`.

We could change it from 

https://github.com/yast/yast-storage-ng/blob/06c72586a788e209bb04dc350a0a2ec66ba542b2/src/lib/y2storage/proposal/settings_generator/base.rb#L79-L81

to 

```ruby
def initialize_settings(settings)
  settings.dup # or sttings.deep_copy
end
```

Reply:
yep, `Yast.deep_copy` goal is to only provide way how old ycp do its object copy. So in general I really do not recommend to do it in new code. It e.g. even do not dup strings as it cause serious slow down of whole yast code and old yast use immutable string, so to fix it Yast.deep_copy do not dup strings and do not handle at all types that is not allowed by yast component system.